CLI: implent update downloading

AdamWill opened the pull request:
This turned out a bit more finicky than I expected, but it does
work. Allows downloading updates by CVE, update ID, and/or build
NVR, accepting comma-separated lists for each. You can use
any combination of the three to download multiple updates at
once.

I couldn't find a better way to do 'one of these three
things must be specified' in click, please let me know if there
is one.

The old CLI added i386 and i586 --arch options for i686 systems,
but that was done back in 2009 referencing an i586 update for
F11; I'm pretty sure we don't have any i386 or i586 builds any
more.

We could add an option to set the arches to download, I guess,
but it makes the code a bit finickier and the old CLI didn't
have it, so I left it out for now.

AdamWill commented:
Additional note: in my own projects I do this:

    # Upstream docs recommend using subprocess32 if possible, but make
    # this optional as it's not available in EPEL.
    try:
        import subprocess32 as subprocess
    except ImportError:
        import subprocess

as the upstream docs recommend using `subprocess32` over `subprocess` if possible. Dunno if you want something like that here.

AdamWill commented:
Also note I kept the consistency between opt name and API query opt name, even though it sucks for humans that most of the opt names are plural but `updateid` is not even though it *does* accept multiple values.

Perhaps this should be fixed in the API - make it accept `updateid` and `updateids` as synonyms?

AdamWill commented on bodhi/cli.py:
```diff
@@ -149,16 +151,45 @@ def comment(update, text, karma, user, password, **kwargs):
+    for item in kwargs.items():
+        if item[1]:
+            expecteds = len(item[1].split(','))
+            resp = client.query(**dict([item]))
```
This is one of those things where I feel like there *must* be a more sensible way to do it, but I'm damned if I could figure it out...:)

ralphbean commented on bodhi/cli.py:
```diff
@@ -149,16 +151,45 @@ def comment(update, text, karma, user, password, **kwargs):
+    # As the query method doesn't let us construct OR queries, we're
+    # gonna run one query for each option that was passed. The syntax
+    # for this is a bit ugly, sorry.
+    for item in kwargs.items():
```
Maybe this could be rephrased for readability a bit.. like:

```python
for attr, value in kwargs.items():
    if not value:
        continue
    expecteds = len(value.split(','))
    resp = client.query(**{attr: value})
    # etc...
```

Names are nicer than indexing ``0`` and ``1`` everywhere.

ralphbean commented on bodhi/cli.py:
```diff
@@ -149,16 +151,45 @@ def comment(update, text, karma, user, password, **kwargs):
+                    # to rewrite it here.
+                    args = ('koji', 'download-build', '--arch=noarch',
+                            '--arch={0}'.format(platform.machine()), build['nvr'])
+                    ret = subprocess.call(args)
```
Maybe if a ``--debug`` or ``--verbose`` option is passed to the command, could you print out the koji command we're running?  :cake:

ralphbean commented:
In general, it looks good to me!  I left some cosmetic comments above.

*caveat emptor* - I've never used the bodhi client 'download' feature before though, so I'm not the best judge of how well it does or doesn't match the functionality of v1.

AdamWill commented:
Yeah, both of those sound like sensible ideas (I think I meant to replace the indexing and just forgot). Thanks!

It's...fairly different from the v1 command. I kinda got the impression the idea wasn't really to mimic the old CLI, since the whole interface (using subcommands) is different anyway.

The old one was only meant to let you specify a single update (it actually enforces this: if the query finds multiple updates, the CLI just says it won't handle that, lists out the updates it found, and bails without downloading any of them). You usually used the old one like this:

`bodhi -D foobar-1.0.fc21`
`bodhi -D FEDORA-2015-123456`

and it downloaded the update you asked for. Now there's a funny wrinkle here. The code in the old CLI looks like this:

            elif opts.download:
                if opts.download.startswith('https://') or \
                   opts.download.startswith('http://'):
                    opts.download = opts.download.split('/')[-1]
                data = bodhi.query(release=opts.release, status=opts.status,
                                   type_=opts.type_, bugs=opts.bugs,
                                   request=opts.request, mine=opts.mine,
                                   limit=opts.limit, package=opts.download)

so it looks like, in theory, you might be able to download packages based on all sorts of other queries, right? Only nope, because a bit further up there's this:

            elif opts.status or opts.bugs or opts.release or opts.type_ or \
                 opts.mine or args:
                [DO SOME QUERY STUFF INSTEAD OF DOWNLOADING]

that `elif` is part of the same `if / else` sequence as the `elif opts.download` at the top of the download stuff. So in fact you can't set `status` or `bugs` or `release` or `type` or `mine` for a download operation, if you try, you just get a query. `request` is elif'ed higher up, so the *only* one you can *actually* set when doing a download is `limit`. That lets you do, I think, just one somewhat funky thing: `bodhi -l1 -D fedfind`, to download the newest update which contains a fedfind package, regardless of any other considerations (e.g. what release it's for or what state it's in). That doesn't seem terribly useful, though, and I'm fairly convinced it wasn't really *intended*. (You can't do `bodhi -l1 -D` to download the very newest update for *any* package, because -D requires an argument).

So it might be nice to make our new client let you do more creative stuff than 'download these specific updates identified by NVR, CVE or update ID', but it's not something the old one did.

There is one way in which this version is a bit less convenient to the user than the old one: the old one let you just pass an NVR or an update ID (or a bare package name, but that's only useful in very odd cases, like the `-l1` case or if there's only a *single* update in Bodhi containing any build of that package) and it would figure out which it was and download it. You didn't have to specify the type of thing you were looking for (as you do in mine).

In Bodhi 1 this was implemented in the server itself - for its query opt `package` it accepted either package names, build NVRs or update IDs, by trying the term as each until it found something. Bodhi 2 does not do this in the server or the API, which is why I had to add it to the client lib `query()` method yesterday (via that PR with the regex'es) to make the old CLI work properly with the new server. Since the server / API don't do it for us in Bodhi 2 the only choices we'd have for doing it that way in the new CLI client are:

1) pass the `package` opt and rely on the BC code I just added to the client lib
2) handle the 'what type of identifier is this' stuff in the CLI code

I don't like 1) for a couple of reasons. First because that code's specifically supposed to be *for backwards compatibility* and presumably we'll want to take it out at some point; new code really ought not to be using hacks we added for BC with old code. Second because the hack wouldn't handle a value like `foobar-1.0-1.fc23,FEDORA-2015-123456,monkeys-3.0-1.fc23` properly; it assumes it's either getting a single value or a list of values all of the same type. This wasn't a problem for the old CLI because it only let you download a single update anyway, but for the new CLI it's a problem because I *do* want to let you download multiple updates at once. In this implementation you can do `bodhi update download --builds foobar-1.0-1.fc23,monkeys-3.0-1.fc23 --updateid FEDORA-2015-123456` and it works fine.

I don't like 2) just because it's kinda rinky-dink special sauce regex stuff in the CLI code; I didn't mind using the regex'es for the BC commit because there's really no great alternative, but for new code I don't really want to be hardcoding assumptions about what identifiers look like into the CLI code, it seems like a bad idea. But if we really want to allow users to just specify different identifiers without telling us what they're meant to be, we could do it, I guess.

The big advantage of my implementation over the original is exactly that you *can* download multiple updates at once.

AdamWill commented:
go home, github, you're drunk.

So I dropped the `item` indexing as suggested. I kinda had second thoughts about the 'log the koji command output' thing, as debug logging seems like something that should be done in an ordered way across the whole command, not just thrown haphazardly into individual subcommands. The way I've been doing it (using argparse) is that there's a top-level `--loglevel` option which you can set to one of the standard log levels, and then logging is done with `logging` - see e.g. https://www.happyassassin.net/cgit/fedfind/tree/fedfind/cli.py . I'm sure there's other ways to do it, but it kinda seems like it should be a co-ordinated approach like that.

ralphbean commented:
> It's...fairly different from the v1 command. I kinda got the impression the idea wasn't really to mimic the old CLI, since the whole interface (using subcommands) is different anyway.

Yeah, that's fair.  We're not all %:100: on how we want to put it together, but I was thinking:

- Try to implement the same functionality so people can do what they used to do, but..
- Rationalize the CLI, since it used to have everything all crammed into one mega command with a gagillion options.  This made it hard for new people (me) to figure out what to do with it.

> So it might be nice to make our new client let you do more creative stuff than 'download these specific updates identified by NVR, CVE or update ID', but it's not something the old one did.

Makes sense.

> The big advantage of my implementation over the original is exactly that you can download multiple updates at once.

**and** that the download code is readable, and not nested in all kinds of ``if this option`` and ``but not this other option`` then ``[DO SOME QUERY STUFF INSTEAD OF DOWNLOADING]``.

:+1: to merge this as-is from me.
